### ISSUE TITLE
Update to CoffeeScript 1.11.0

### ISSUE THREAD

**Issue by alangpierce** [decaffeinate/decaffeinate#427]
CoffeeScript released a new version, 1.11.0: https://github.com/jashkenas/coffeescript/pull/4315 . The main new feature is import/export syntax. Most likely, there will eventually be people with CoffeeScript 1.11.0 code who want to use decaffeinate, and current decaffeinate only supports 1.10.0 code.

There are a few things that need to be done to move to the new version:
* Update the [decaffeinate fork of CoffeeScript](https://github.com/decaffeinate/coffeescript) to be based on the new version and re-apply the one patch that didn't make it into 1.11.0.
* Update decaffeinate-parser to handle the new AST format. There were some additions and some breaking changes to CoffeeScript AST format, so this will probably be the biggest change.
* Make decaffeinate handle the new import/export syntax. Since the CoffeeScript syntax is heavily based on the ES2015 syntax, hopefully this won't be so bad.

For the near future, anyone wanting to use decaffeinate likely has 1.10.0 code. For example, in the main codebase I'm work with, we won't be updating to CS 1.11.0 since we're already using JavaScript for new code and want to move off of CoffeeScript. So this task doesn't seem urgent, but would be nice to get done eventually. Feel free to react or comment if you have CoffeeScript 1.11.0 code that you would like to convert.

**Comment by wavebeem** [comment]
For what it's worth, we have a large project at work that's now using a lot of ES6 including ES6 modules, but has a lot of CoffeeScript code, and we would *love* to use this project but can't until it supports ES6 modules.

**Comment by alangpierce** [comment]
@wavebeem thanks for the feedback. Do you mean that you've updated to CoffeeScript 1.11.0 (or later) and started moving CoffeeScript code to use import/export syntax? If you have commonjs CoffeeScript, decaffeinate already does output ES6 modules; it generally tries to rewrite `require` and `module.exports` into `import` and `export`. [Here's a repl link with some examples.](http://decaffeinate-project.org/repl/#?evaluate=false&stage=full&code=a%20%3D%20require('b')%0A%7Bc%7D%20%3D%20require('d')%0Ae%20%3D%20require('f').g%0A%0Amodule.exports%20%3D%20'hello%20world'%3B%0Amodule.exports%20%3D%20%7B%0A%20%20h%3A%20i%0A%7D%0Amodule.exports.j%20%3D%20k%0A)

It's certainly a bit fragile, though, particularly since named imports/exports don't work exactly like module fields. The [bulk-decaffeinate](https://github.com/alangpierce/bulk-decaffeinate) project has an optional fix-imports follow-up step that can correct for this, so that's what I'd recommend.

Does that address your issue? Certainly one migration path I've thought about for my codebase is to migrate to the new CS 1.11.0 import/export system, then run it through decaffeinate (after fixing this issue), but I think that's probably more trouble than it's worth, and it's easier to just move from commonjs CS 1.10.0 directly to JS.

**Comment by wavebeem** [comment]
Currently we're using code like this:

```coffeescript

`import Ember from 'ember'`
`import {foo, bar} from 'baz'`

class CoffeeScript
  whatever: =>
    @something

`export default Coffeescript`
```

But we plan to upgrade to Coffeescript 1.11.0 and remove the backticks so the ES6 modules work without the inline-JS hack.

Currently the backticks cause an error to be thrown, so I tried removing them, but that didn't work either.

Our use case is an old Ember app with mixed CoffeeScript and JS, which we're porting to a new Ember CLI app using Babel and CoffeeScript.

**Comment by wavebeem** [comment]
So yeah, the move to ES6 modules inside our CoffeeScript was already done. It never used CommonJS originally, it just used globals on the page. So we don't have any CommonJS version to roll back to and try decaffeinate on it.

**Comment by alangpierce** [comment]
Interesting, I hadn't seen backticks used to get imports in CoffeeScript, that makes sense. decaffeinate generally has support for backtick syntax, so I think the easiest path is to make that work.

I took a look and it looks like the error is actually just a silly parsing bug in decaffeinate-parser. Here's a [repl link](http://decaffeinate-project.org/repl/#?evaluate=false&stage=full&code=%60import%20Ember%20from%20'ember'%20%60%0A%60import%20%7Bfoo%2C%20bar%7D%20from%20'baz'%20%60%0A%0Aclass%20CoffeeScript%0A%20%20whatever%3A%20%3D%3E%0A%20%20%20%20%40something%0A%0A%60export%20default%20Coffeescript%60) that works around the issue by adding a space after the close-quote. That output is what you want, right?

I can file a bug for that parser crash, and I bet it won't be hard to fix soon.

**Comment by wavebeem** [comment]
Ah nice, that looks like it'll work. I'll just do a search and replace on our repo and see how it goes. Thanks.

Also, thanks for the cool project, I'm really hoping we can ditch our Coffee at work with this :v:

**Comment by alangpierce** [comment]
Thanks, glad to hear it's helpful. 😄 

The parser bug just got fixed with https://github.com/decaffeinate/decaffeinate-parser/pull/89, and it was just released as decaffeinate-parser 4.0.1. The repl isn't updated yet because it only gets updated on changes to the core decaffeinate project, but if you reinstall decaffeinate (or otherwise pull in the latest dependencies) the crash should be fixed, so you shouldn't need the space workaround.

**Comment by dkryaklin** [comment]
@alangpierce Hey, what about support import/export without backticks?
E.g.
`import React from 'react'
import {connect} from 'react-redux'

class App extends React.Component
  @displayName: 'App'

export {App}
`

**Comment by alangpierce** [comment]
@dkryaklin yep, that's in the scope of this task. CoffeeScript 1.11.0 added that syntax, and decaffeinate currently uses the CoffeeScript 1.10.0 parser, so import/export syntax isn't supported yet, but updating to the new parser would make it work.

**Comment by stuta** [comment]
Will the import/export syntax be supported? In some fork or options?

**Comment by meenie** [comment]
Any update on this? It's preventing our team from converting our massive CoffeeScript codebase to JavaScript.

**Comment by alangpierce** [comment]
@meenie I haven't been working on it since my CoffeeScript codebase didn't have any import statements, but I can try to get it working.

Which features are you using? Are you using CS2 or just a later version of CS1? (CS2 has some differences in behavior, both in the syntax and in the semantics, so it would need a special mode.)

Here are the newer CS features that I'm aware of that aren't supported by decaffeinate:
* `import`/`export`.
* `for`...`from`.
* Tagged template literals.
* Async functions
* Object rest/spread
* JSX

**Comment by meenie** [comment]
We are using `1.11.1` and the biggest hangup is we are using `import` / `export`. It wouldn't be that big of an issue but we do `export`'s like this:
```
export default Ember.Component.extend ResizeHandler,
  ...
```

Otherwise we would just use backticks.

Thanks for your time!

**Comment by jhpratt** [comment]
Just ran across this repo, and definitely like it over plain CS. I'm currently using CS@2.1.1, and the main thing preventing me from switching compilers right now is `import`/`export` and `async`/`await`. If those are resolved, I'd definitely switch.

**Comment by eventualbuddha** [comment]
@jhpratt decaffeinate really isn't intended to be used as a CoffeeScript compiler in the sense that you write and maintain CoffeeScript code and use it to convert to JavaScript for the browser. Its goal is to help you *replace* your CoffeeScript code with JavaScript code one time, and from then on you write and maintain JavaScript code.

In theory you can use decaffeinate in the way you describe, but it would be slower and isn't a drop-in replacement for the official CoffeeScript compiler.

**Comment by boris-petrov** [comment]
@eventualbuddha - what @jhpratt is using `decaffeinate` for is indeed strange however I agree with him that `import`/`export` and `async`/`await` are missing. That also stops us from converting all our CoffeeScript 2 code to JavaScript (and then on to TypeScript).

**Comment by jhpratt** [comment]
Hmm, I hadn't put thought as to the fact that it's intended to be a one-time converter. Regardless of my (admittedly odd) use case, I'm sure others could make use of `import`/`export` and `async`/`await` with regards to CS2.

**Comment by benc** [comment]
If you have a codebase with CS2 and want to migrate to TypeScript, an alternative solution is to use the coffee compiler. It will output ES2015 syntax.

`coffee --compile --bare --no-header foo.coffee`

I've used the above command to migrate several codebases to Typescript. There are some constructs that decaffeinate handles more elegantly though, so I've used decaffeinate on some snippets.

**Comment by boris-petrov** [comment]
@benc - thanks for the idea! Could you elaborate which constructs `decaffeinate` handles better?

**Comment by benc** [comment]
By better I mean "more readable", at least in my opinion.

I've found for example that usage of the [existential operator](http://coffeescript.org/#existential-operator) has a more readable output using `decaffeinate`. ([it does not exist yet in JS/TS](https://github.com/tc39/proposal-optional-chaining))

The bulk of the code however converted quite nicely.

**Comment by alangpierce** [comment]
Hi, sorry for letting this sit for a while. Looks like there's definitely demand for running decaffeinate on newer CS code, so I'll try to post a technical plan here pretty soon. I think I should have bandwidth to get to it in the next two weeks or so. It seems probably easiest for me and/or @eventualbuddha to work on this, but also happy to try to coordinate if anyone else wants to contribute (especially if it ends up taking a while).

I'll update the issue title from "Update to CoffeeScript 1.11.0" to the more general "Update to latest CoffeeScript". Given that CS2 is a breaking change, I believe that we'll need two modes:
* CS1 mode will be based on a fork of the CoffeeScript 1.12.7 parser. Like decaffeinate today, it will follow CS1 semantics, e.g. using default values when the arg is either `null` or `undefined`. My plan is for this to be an intermediate milestone, and will probably satisfy many use cases (e.g. it will support import and export). I think this should continue to be the default, since I think most people wanting to move off of CoffeeScript have CS1 codebases.
* CS2 mode will be based on a fork of the CoffeeScript 2.2.0 parser. It will follow the semantic behavior of CS2, which generally means disabling certain defensive transformations in favor of the more straightforward ones.

I believe we'll need a different CS parser for each mode because CS2 had the breaking change of removing bare `super`, although it's possible that the CS2 parser actually handles it. I've also seen a lot of fragility around whitespace handling in the CS parser, so using different parsers feels safest to me.


If anyone knows of public CS2 projects (or CS1 projects using newer syntax) that can be used as test cases, it would be great to hear! Probably my biggest worry is overlooking something because I don't have a nice CS2 codebase to try everything out on. Feel free to comment here or send a PR to the [decaffeinate-examples](https://github.com/decaffeinate/decaffeinate-examples) repo.

@boris-petrov a while back I wrote up a blog post on decaffeinate, which in part talks about how it compares with the CoffeeScript compiler. https://benchling.engineering/from-200k-lines-of-coffeescript-to-zero-making-decaffeinate-super-stable-4de0ca68d9bc Generally I'd say that decaffeinate gives significantly better results than the CS1 compiler and somewhat better results than the CS2 compiler. Another high-level approach if you'd rather not wait for newer CS support is to modify each file to work with decaffeinate, run it through decaffeinate, then modify it back.

**Comment by boris-petrov** [comment]
@alangpierce - thanks for the blog post and the intention of helping `decaffeinate` support CS2! As for a big CS2 project - I'm not sure, but perhaps the CoffeeScript compiler itself is written in CS2? Otherwise - the differences between 1 and 2 are not that major at all. 2 added async/await support, splatting here and there... and that's pretty much all I can think of that we are actually using.

Thanks again for the support!

**Comment by alangpierce** [comment]
Yep, you're right, the CoffeeScript 1 compiler and its test suite were certainly good test cases for decaffeinate, so the latest code should hopefully be a good CS2 test case.

And yeah, CS2 is certainly a pretty minor change to the language itself. It's possible that the CS2 parser (or a simple modification to it) could handle CS1 including all edge cases, so certainly seems worth trying. 🤞

**Comment by alangpierce** [comment]
Here’s my plan for the steps that will be needed to get decaffeinate working with CS 1.12.7 and CS2. It will also have a side-effect of adding JSX support (at least when using the CS2 parser). I'll check things off as I get to them, and some may spin off into separate tasks.

## Fix build issues
We’ll want working tests for all projects to avoid regressions due to the parser change.

- [ ] Atom: Failure when installing old packages through apm. I already sent a PR to fix this, and will need to pull in the new apm version once it's published.
- [ ] Codecombat: It looks like they recently started adding .vue files with inline CoffeeScript, so decaffeinate doesn’t convert them and they fail to compile because the patch removes CoffeeScript from the build system. Possibly an easy workaround is to pin to a fixed Codecombat version. 
- [ ] Autoprefixer: recent test failure probably due to external dependency that changed. A quick solution is maybe to just disable that one test.

## Update to CS 1.12.7 with current feature set
- [ ] Look through the 10 or so patches in the CS fork and apply any that aren’t applied in 1.12.7
- [ ] Change version scheme to properly handle numerical order
- [ ] Update TypeScript typings for the new AST format
- [ ] Update decaffeinate-parser to handle changes in AST format (without adding new nodes)

## Add support for new syntax in CS 1.12.7
New syntax:
* `import` and `export`
* Tagged template literals
* `for...from`
* Triple backticks for embedded JS blocks
* Escaping within JS backtick expressions
* `@index` assignees in `for` expressions
* Bare `yield`
* Make sure it’s ok to drop support for `a && = b`
* Make sure `”””` indentation stripping changes do not cause problems

Tasks:
- [ ] Update coffee-lex to support new tokens. (`from` will still be an identifier)
- [ ] Add new node types to decaffeinate-parser and transform from CS AST
- [ ] Add decaffeinate patchers
- [ ] Update decaffeinate-examples to use CS 1.12.7 test suite

## Figure out how to properly use the CS2 parser
This is the biggest option question. CS2 drops support for some CS1 syntax that would be nice to support in decaffeinate:
* Bare `super`
* Bound generator functions
* `extends` binary operator
* Changes in `<` and `>` sometimes treated as JSX

It seems unlikely that I’ll be able to modify the CS2 parser to still support this syntax. For example, bare `super` now just gives an “indentation error”; it’s implicitly disallowed to due a change in the grammar, rather than explicitly disallowed.

Here are the options I’m considering, from most to least desirable:
* Define a single TypeScript type signature that encompasses the CS1 and CS2 AST, and make decaffeinate-parser run on only that. This relies in the runtime objects being compatible.
* Make minor modifications to the CS1 and CS2 forks (ideally just the CS1 fork) so that the AST formats are compatible.
* Define separate type signatures for CS1 and CS2, and write a transform from CS1 AST to CS2 AST, which runs at the start of decaffeinate-parser. Add any necessary flexibility in the CS2 AST to encompass both. Then decaffeinate-parser just operated on a CS2 AST.
* Publish two versions of decaffeinate-parser, one which operates on the CS1 AST and one which operates on the CS2 AST. Both emit the same decaffeinate-parser AST format.

## Update decaffeinate to optionally use the CS2 parser
- [ ] Add a flag to decaffeinate and decaffeinate-parser to use the CS2 parser.
- [ ] Implement any changes needed in decaffeinate-parser to produce an AST equivalent to before.

## Add support for new syntax in CS2
New syntax:
* `async`/`await`
* JSX
* Object rest/spread
* Inline `###` comments
* `...` on the left side of an expression.
* Computed property keys in objects

Tasks:
- [ ] Update coffee-lex to support new tokens. JSX will likely be the most difficult aspect of this.
- [ ] Add new nodes to decaffeinate-parser.
- [ ] Add new patchers and update existing patchers to handle new syntax options.

## Add support for new semantics in CS2
When using the CS2 parser, decaffeinate should change the behavior of certain transforms:
- [ ] Transform default values to JS default values, including in destructuring.
- [ ] Don’t emit `Array.from` for array destructuring.
- [ ] Bound subclass methods are set after the call to `super`
- [ ] Backtick expressions in class bodies should be placed in the JS class body.

**Comment by alangpierce** [comment]
@dkryaklin @stuta @meenie @boris-petrov

As of decaffeinate 3.5.0, decaffeinate should work on all CoffeeScript 1.12.7 code. That means it'll understand and transform these features:
* `import` and `export`
* Tagged template literals
* `for...from`
* Triple backticks for inline JS

(Plus various other minor things; see the CoffeeScript release notes for full details.)

decaffeinate does *not* yet have full CS2 support, but even with a CS2 codebase this may be enough to make decaffeinate usable.

Let me know if you see any issues! There are plenty of tests, but it hasn't been "battle-tested" nearly as much as the other parts of decaffeinate.

Upgrading the CoffeeScript test suite exposed some minor correctness issues (details around unicode in regexes and escaping within inline JS), which I'm planning to fix when I get a chance, but other than that, everything should hopefully just work.

**Comment by meenie** [comment]
@alangpierce It worked for us :). Thanks!

**Comment by jhpratt** [comment]
@alangpierce Found a bug. In both `import` and `export`, anything between braces doesn't need a comma, but decaffeinate throws an error.

Simple test cases:

    import {
      foo
      bar
    }

and

    export {
      foo
      bar
    }

**Comment by alangpierce** [comment]
@jhpratt Thanks, I just filed as https://github.com/decaffeinate/decaffeinate/issues/1244 . That and full CS2 support should be doable reasonably soon, although I've been pretty busy recently.

**Comment by boris-petrov** [comment]
@alangpierce - sounds great, I'll happily test it on our CS2 codebase (which is pretty large) when it's done! :) Thanks!

**Comment by alangpierce** [comment]
@boris-petrov as of decaffeinate 4.8.1, CS2 support should be pretty much ready! You can use the `--use-cs2` flag to enable it. There are still some remaining little correctness issues (hopefully obscure enough that you'd be unlikely to hit them), so I'm leaving this issue open for now, but it should support JSX, async/await, object rest/spread, computed keys, and all of the other new syntax, and most of the small semantics changes have been incorporated.

**Comment by meenie** [comment]
Well done, @alangpierce!

**Comment by boris-petrov** [comment]
@alangpierce - thanks a bunch, you're the best! I have a problem with `await` statements though. Using `decaffeinate --use-cs2 .` with this code:

```coffeescript
export default Component.extend
  actions:
    selectProperty: (property) ->
      @_resetValues()

      # TODO: something
      propertyType = await property.get('propertyType')
      if property.get('isArray')
        filters = [
```

And that fails with:

```
AddVariableDeclarationsStage failed to parse: await is a reserved word (53:21)
  51 | 
  52 |       // TODO: something
> 53 |       propertyType = await property.get('propertyType');
     |                      ^
  54 |       if (property.get('isArray')) {
  55 |         filters = [
```

Am I doing something wrong?

**Comment by alangpierce** [comment]
@boris-petrov nope, looks like a bug! "await is a reserved word" is the error you get in JS when you forget to put `async` on a function and use `await` inside of it, and looks like decaffeinate isn't adding `async` to the method when it should. It already knows to add `*` for generators, so I think it just needs to do the same for async functions. I just filed #1285 to track.

**Comment by alangpierce** [comment]
@boris-petrov should be fixed in decaffeinate 4.8.2. Let me know if that fixes things or you run into anything else!

**Comment by boris-petrov** [comment]
@alangpierce - thanks for the quick fix! All of our project seems to be translated without error. I don't have the time now but in a week or two, I'll definitely be trying it out and migrating to that code after that. :) Thanks a bunch, I'll notify you when that happens!

**Comment by fent** [comment]
any chance we could get this in bulk-decaffeinate?

**Comment by eventualbuddha** [comment]
@fent bulk-decaffeinate just shells out to decaffeinate, so if you have latest decaffeinate then bulk-decaffeinate should work.
